iphone/Maps/Core/Location/MWMLocationManager.mm

```diff
@@ -469,43 +484,24 @@ - (void)setStarted:(BOOL)started
   }
 }
 
-- (BOOL)start
+- (void)requestPermissionIfNeededFor:(CLLocationManager *)manager
 {
-  MWMVoidBlock doStart = ^{
-    LOG(LINFO, ("startUpdatingLocation"));
-
-    CLLocationManager * locationManager = self.locationManager;
-    if ([locationManager respondsToSelector:@selector(requestWhenInUseAuthorization)])
-      [locationManager requestWhenInUseAuthorization];
-
-    [locationManager startUpdatingLocation];
-
-    setPermissionRequested();
+  [manager requestWhenInUseAuthorization];
+}
 
-    if ([CLLocationManager headingAvailable])
-      [locationManager startUpdatingHeading];
-  };
+- (void)startUpdatingLocationFor:(CLLocationManager *)manager
+{
+  LOG(LINFO, ("startUpdatingLocation"));
+  
+  [manager startUpdatingLocation];
+  if ([CLLocationManager headingAvailable])
+    [manager startUpdatingHeading];
+}
 
-  if ([CLLocationManager locationServicesEnabled])
-  {
-    switch ([CLLocationManager authorizationStatus])
-    {
-    case kCLAuthorizationStatusAuthorizedWhenInUse:
-    case kCLAuthorizationStatusAuthorizedAlways:
-    case kCLAuthorizationStatusNotDetermined:
-        doStart();
-        return YES;
-    case kCLAuthorizationStatusRestricted:
-    case kCLAuthorizationStatusDenied:
-        [self processLocationStatus:MWMLocationStatusDenied];
-        break;
-    }
-  }
-  else
-  {
-    [self processLocationStatus:MWMLocationStatusGPSIsOff];
-  }
-  return NO;
+- (BOOL)start
+{
+  [self locationManager];
```

Comment:
What does it do? Is it required?

Reply:
This call is used only to intitalize the `locationManager` instance (once on the first access).
There are several places where the location manager may be called when the app is starting: in `start`, in geomode `setGeoMode` and other places (but usually in the `start`). Without this calling delegate will not work.
I can add the comment describing why this getter is calling.

```swift
- (CLLocationManager *)locationManager
{
  if (!_locationManager)
  {
    _locationManager = [[CLLocationManager alloc] init];
    _locationManager.delegate = self;
    [self refreshGeoModeSettings];
    _locationManager.pausesLocationUpdatesAutomatically = YES;
    _locationManager.headingFilter = 3.0;
  }
  return _locationManager;
}

```

Reply:
Would it be better to initialize this property in MWMLocationManager's constructor?

Will this work?

```
// Create and initialize CLLocationManager object.
(void)self.locationManager;
```

Reply:
@biodranik I don't think so, because during the init process of the CLLocationManager method `[self refreshGeoModeSettings];` is called. And for this call the MWMLocationManager instance should exist. So it can prodece unexpected behavior.

Reply:
Is it a cyclic dependency on calling self.locationManager from refreshGeoModeSettings that is called from self.locationManager?

Would it be better to extract the static function to get settings and use it in the self.locationManager?

Reply:
This code smells a bit, but it works because of location manager initialize once and getter returns ivar _locationManager. On second call of Self.licationManager will not call refreshGeo mode.

I'll try to refactor it a little bit